### PR TITLE
Optimize GetPropsFromDict: use tags for conversion

### PR DESCRIPTION
**Pull request by bp-kelley**
This is a test for optimizing GetPropsAsDict, it should be >10x faster in practice if it doesn't segfault the tests.

### CODE REVIEW

File: Code/GraphMol/Wrap/props.hpp [hunk 1]
```diff
@@ -76,42 +77,67 @@ bool AddToDict(const U &ob, boost::python::dict &dict, const std::string &key) {
...
-    }
-    if (AddToDict<std::vector<double>>(obj, dict, keys[i])) {
-      continue;
-    }
-    if (AddToDict<std::vector<std::string>>(obj, dict, keys[i])) {
-      continue;
-    }
-    if (AddToDict<std::string>(obj, dict, keys[i])) {
+  for (auto &rdvalue: data) {
+    if (std::find(keys.begin(), keys.end(), rdvalue.key) == keys.end())
```
Comment (greglandrum) [inline Code/GraphMol/Wrap/props.hpp:87]: Out of curiosity: how can it happen that one of the keys from rdvalue is not in `keys`?

File: Code/GraphMol/Wrap/props.hpp [hunk 2]
```diff
@@ -76,42 +77,67 @@ bool AddToDict(const U &ob, boost::python::dict &dict, const std::string &key) {
...
+	dict[rdvalue.key] = from_rdvalue<int>(rdvalue.val);
+	break;
+      case RDTypeTag::DoubleTag:
+	dict[rdvalue.key] = from_rdvalue<double>(rdvalue.val);
+	break;
+      case RDTypeTag::StringTag:
+	if (autoConvert) {
+	  // Auto convert strings to ints and double if possible
+	  if(AddToDict<int>(obj, dict, rdvalue.key)) break;
+	  else if(AddToDict<double>(obj, dict, rdvalue.key)) break;
```
Comment (greglandrum) [inline Code/GraphMol/Wrap/props.hpp:102]:
```suggestion
	  if(AddToDict<int>(obj, dict, rdvalue.key)) {
	    break;
	  } else if(AddToDict<double>(obj, dict, rdvalue.key)) { 
	    break;
	  }
```

File: Code/GraphMol/Wrap/props.hpp [hunk 3]
```diff
@@ -76,42 +77,67 @@ bool AddToDict(const U &ob, boost::python::dict &dict, const std::string &key) {
...
+      case RDTypeTag::VecIntTag:
+	dict[rdvalue.key] = from_rdvalue<std::vector<int>>(rdvalue.val);
+	break;
+      case RDTypeTag::VecUnsignedIntTag:
+	dict[rdvalue.key] = from_rdvalue<std::vector<unsigned int>>(rdvalue.val);
+	break;
+      case RDTypeTag::VecStringTag:
+	dict[rdvalue.key] = from_rdvalue<std::vector<std::string>>(rdvalue.val);
+	break;
+      }
```
Comment (greglandrum) [inline Code/GraphMol/Wrap/props.hpp:133]: Maybe add a `default` here with with an `UNDER_CONSTRUCTION("not implemented")` invariant to catch the situation where we add a new type tag and forget to update this file?

File: Code/GraphMol/Wrap/Atom.cpp [hunk 4]
```diff
@@ -388,7 +388,8 @@ struct atom_wrapper {
 
         .def("GetPropsAsDict", GetPropsAsDict<Atom>,
              (python::arg("self"), python::arg("includePrivate") = true,
-              python::arg("includeComputed") = true),
+              python::arg("includeComputed") = true,
+	      python::arg("autoConvert") = true),
```
Comment (greglandrum) [inline Code/GraphMol/Wrap/Atom.cpp:392]: I think the name of this argument should be `autoConvertStrings` to make it more explicit when the autoconversion actually happens.

**Review by greglandrum — CHANGES_REQUESTED** [review]
This makes soooo much sense.
small changes requested.

File: Code/GraphMol/Wrap/props.hpp (hunk 3 above)
Comment (bp-kelley) [reply Code/GraphMol/Wrap/props.hpp:136]: Will do.

**Comment by bp-kelley** [comment]
@greglandrum I was thinking about adding a GetProp(propane, autoConvert=True) or a GetPyProp(propane) in order to get a single value and not having to return the dictionary.

Preferences?

**Comment by greglandrum** [comment]
> @greglandrum I was thinking about adding a GetProp(propane, autoConvert=True) or a GetPyProp(propane) in order to get a single value and not having to return the dictionary.
> 
> Preferences?

Nice idea! I like the `autoConvert=True` form better.

**Review by greglandrum — APPROVED** [review]
LGTM

**Comment by greglandrum** [comment]
> > @greglandrum I was thinking about adding a GetProp(propane, autoConvert=True) or a GetPyProp(propane) in order to get a single value and not having to return the dictionary.
> > Preferences?
> 
> Nice idea! I like the `autoConvert=True` form better.

I think that's a separate PR though. Or would you prefer to do it in this one?

**Comment by bp-kelley** [comment]
Let me see how the code refactors first, otherwise I'll make it a second PR.

**Comment by bp-kelley** [comment]
I added the GetProp enhancement, it defaults to False for now.  I added the tests to `testPropertyLists.py` mainly because it pains me to add anything more to rough_test.py :). I had to add C++ test code as well to actually set the properties.  This tests

Mol
Atom
Bond
Conformer

but not substance group.

**Comment by greglandrum** [comment]
> I added the tests to `testPropertyLists.py` mainly because it pains me to add anything more to rough_test.py :)

Adding new test files is almost never wrong! :-)

**Review by greglandrum — CHANGES_REQUESTED** [review]
I think the C++ testing function should be renamed in python, but otherwise this looks good.

**Review by greglandrum — APPROVED** [review]
LGTM
will merge once the CI passes

**Comment by bp-kelley** [comment]
@greglandrum it has the same problem I saw before, we do this in other places, any idea why adding the underscore doesn't work here?

**Comment by bp-kelley** [comment]
@greglandrum I found the issue, I need to include private vars from their libs not from Chem.  I.e. rdkit.Chem.rdmolops

**Pull request merged**